File: docs-source/docs/modules/microservices-tutorial/examples/05-shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityProjectionHandler.scala
// tag::handler[]
package shopping.cart

import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import scala.util.Success

import akka.Done
import akka.actor.typed.ActorSystem
import akka.projection.eventsourced.EventEnvelope
import akka.projection.scaladsl.Handler
import org.slf4j.LoggerFactory

class ItemPopularityProjectionHandler(
    tag: String,
    system: ActorSystem[_],
    repo: ItemPopularityRepository)
    extends Handler[EventEnvelope[ShoppingCart.Event]]() { // <1>

  private val log = LoggerFactory.getLogger(getClass)
  private implicit val ec: ExecutionContext =
    system.executionContext

  override def process(
      envelope: EventEnvelope[ShoppingCart.Event])
      : Future[Done] = { // <2>
    val processed = envelope.event match { // <3>
      case ShoppingCart.ItemAdded(_, itemId, quantity) =>
        repo.update(itemId, quantity)

      // end::handler[]
      case ShoppingCart.ItemQuantityAdjusted(
            _,
            itemId,
            newQuantity,
            oldQuantity) =>
        repo.update(itemId, newQuantity - oldQuantity)

      case ShoppingCart.ItemRemoved(
            _,
            itemId,
            oldQuantity) =>
        repo.update(itemId, 0 - oldQuantity)
      // tag::handler[]

      case _: ShoppingCart.CheckedOut =>
        Future.successful(Done)
    }
    processed.onComplete {
      case Success(_) => logItemCount(envelope.event)
      case _          => ()
    }
    processed
  }

  private def logItemCount(
      event: ShoppingCart.Event): Unit =
    event match {
      case itemEvent: ShoppingCart.ItemEvent =>
        val itemId = itemEvent.itemId
        repo.getItem(itemId).foreach {
          case Some(count) =>
            log.info(
              "ItemPopularityProjectionHandler({}) item popularity for '{}': [{}]",
              tag,
              itemId,
              count)
          case None =>
            log.info(
              "ItemPopularityProjectionHandler({}) item popularity for '{}': [0]",
              tag,
              itemId)
        }
      case _ => ()
    }

}
// end::handler[]

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-java/src/main/java/shopping/cart/ItemPopularityProjectionHandler.java
// tag::handler[]
package shopping.cart;

import akka.projection.eventsourced.EventEnvelope;
import akka.projection.jdbc.javadsl.JdbcHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.persistence.EntityManager;
import javax.persistence.NoResultException;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;

public final class ItemPopularityProjectionHandler
    extends JdbcHandler<EventEnvelope<ShoppingCart.Event>, HibernateJdbcSession> { // <1>
  private final Logger logger = LoggerFactory.getLogger(getClass());
  private final String tag;
  private final ItemPopularityRepository repo;

  public ItemPopularityProjectionHandler(String tag, ItemPopularityRepository repo) {
    this.tag = tag;
    this.repo = repo;
  }

  @Override
  public void process(
      HibernateJdbcSession session, EventEnvelope<ShoppingCart.Event> envelope) { // <2>
    ShoppingCart.Event event = envelope.event();

    if (event instanceof ShoppingCart.ItemAdded) { // <3>
      ShoppingCart.ItemAdded added = (ShoppingCart.ItemAdded) event;
      String itemId = added.itemId;

      // FIXME testing entityManager operations (within same transaction)
      EntityManager entityManager = session.entityManager();
      CriteriaBuilder builder = entityManager.getCriteriaBuilder();
      CriteriaQuery<ItemPopularity> criteria = builder.createQuery(ItemPopularity.class);
      Root<ItemPopularity> root = criteria.from(ItemPopularity.class);
      criteria.select(root).where(builder.equal(root.get("itemId"), itemId + "_a"));
      try {
        ItemPopularity result = entityManager.createQuery(criteria).getSingleResult();
        logger.info("Existing {}: {}", result.getItemId(), result.getCount());
          ItemPopularity updatedItemPop = result.changeCount(added.quantity);
          entityManager.merge(updatedItemPop);
      } catch (NoResultException e) {
        logger.info("New {}", itemId + "_a");
        ItemPopularity newItemPop = new ItemPopularity(itemId + "_a", 0, added.quantity);
        entityManager.persist(newItemPop);
      }

      ItemPopularity existingItemPop =
          repo.findById(itemId).orElseGet(() -> new ItemPopularity(itemId, 0, 0));
      ItemPopularity updatedItemPop = existingItemPop.changeCount(added.quantity);
      repo.save(updatedItemPop);
      logger.info(
          "ItemPopularityProjectionHandler({}) item popularity for '{}': [{}]",
          this.tag,
          itemId,
          updatedItemPop.getCount());

      // FIXME tested that this will rollback both above operations and the offset storage
//      throw new RuntimeException("Simulated exc in projection handler");
      // end::handler[]
    } else if (event instanceof ShoppingCart.ItemQuantityAdjusted) {
      ShoppingCart.ItemQuantityAdjusted adjusted = (ShoppingCart.ItemQuantityAdjusted) event;
      String itemId = adjusted.itemId;
      ItemPopularity existingItemPop =
          repo.findById(itemId).orElseGet(() -> new ItemPopularity(itemId, 0, 0));
      ItemPopularity updatedItemPop =
          existingItemPop.changeCount(adjusted.newQuantity - adjusted.oldQuantity);
      repo.save(updatedItemPop);
    } else if (event instanceof ShoppingCart.ItemRemoved) {
      ShoppingCart.ItemRemoved removed = (ShoppingCart.ItemRemoved) event;
      String itemId = removed.itemId;
      ItemPopularity existingItemPop =
          repo.findById(itemId).orElseGet(() -> new ItemPopularity(itemId, 0, 0));
      ItemPopularity updatedItemPop = existingItemPop.changeCount(-removed.oldQuantity);
      repo.save(updatedItemPop);
      // tag::handler[]
    } else {
      // skip all other events, such as `CheckedOut`
    }
  }
}
// end::handler[]
